tidy3d/components/geometry.py

```diff
@@ -3227,8 +3230,212 @@ def _surface_area(self, bounds: Bound) -> float:
         return area
 
 
+class CustomSurfaceMeshGeometry(Geometry, ABC):
+    """Custom surface geometry given by a triangle mesh, as in the STL file format.
+
+    Example
+    -------
+    >>> import stl
+    >>> stl_mesh = stl.mesh.Mesh(np.zeros(4), dtype=stl.mesh.Mesh.dtype)
+    >>> # tetrahedron
+    >>> stl_mesh.vectors = [
+    >>>     [[1,0,0],[0,1,0],[0,0,1]],
+    >>>     [[0,0,0],[0,0,1],[0,1,0]],
+    >>>     [[0,0,0],[1,0,0],[0,0,1]],
+    >>>     [[0,0,0],[0,1,0],[1,0,0]]
+    >>> ]
+    >>> stl_geom = CustomSurfaceMeshGeometry.from_stl(stl_mesh)
+    """
+
+    mesh_dataset: Optional[SurfaceMeshDataset] = pydantic.Field(
```

Comment:
When is it used as `Optional`?

Reply:
If this is not `Optional`, then we fail the four IO tests concerned with writing to JSON and not HDF5.

Reply:
@weiliangjin2021 when a component containing a dataset is saved to json file, the data arrays are not written directly to the file, instead just a string is put there indicating the type of data array. Then, when the component is load from that file, the component is stored as None. This is why it needs to be optional. We made it this way so the web interface can still use the json file even if data is present.